src/core/queriesObserver.ts

```diff
@@ -64,78 +64,120 @@ export class QueriesObserver extends Subscribable<QueriesObserverListener> {
   }
 
   getOptimisticResult(queries: QueryObserverOptions[]): QueryObserverResult[] {
-    return queries.map((options, index) => {
-      const defaultedOptions = this.client.defaultQueryObserverOptions(options)
-      return this.getObserver(defaultedOptions, index).getOptimisticResult(
-        defaultedOptions
-      )
+    return this.getUpdatedObservers(queries).newResult
+  }
+
+  private getUpdatedObservers(
+    queries: QueryObserverOptions[],
+    notifyOptions?: NotifyOptions
+  ): QueryObserverUpdate {
+    const prevObservers = this.observers
+    const prevObserversMap = this.observersMap
+    const newResult: QueryObserverResult[] = []
+    const newObservers: QueryObserver[] = []
+    const newObserversMap: Record<string, QueryObserver> = {}
+
+    const defaultedQueryOptions = queries.map(o =>
+      this.client.defaultQueryObserverOptions(o)
+    )
+    const matchingObservers = defaultedQueryOptions
+      .map(o => {
+        if (o.queryHash == null) {
```

Comment:
I'd appreciate it if we wouldn't use one-letter abbreviations. Using `option` here instead of `o` would't hurt I believe, especially because further down, `o` is also used as abbreviation for `observer`, which makes this quite confusing ....